### PR TITLE
Editor fast follow

### LINKED ISSUE #575: Markdown editor fast fixes - 0.5.0 beta
This issue can be appended to include various bugs that are found in the new editor. When fast fixes are complete, any remaining issues here can be converted to standalone issues.

#### Inline node processing issues
Inline nodes need to be processed alongside surrounding text, and be subject to surrounding marks, while still having their children processed as required.
- [x] inline images become text in the visual editor
- [x] markdown images within links used to display as images, now they display as text
- [x] marks around links cause the links to become text in the visual editor
- [x] marks applied to inline code, e.g. \*\*\`test\`\*\*, break in the visual editor

#### Escaping issues
- [x] when converting markdown to Remark, don't parse markdown entities within raw html
- [x] when converting Slate to Remark, escape all markdown entities except those in raw html
- [x] markdown images broken by unintelligent escaping because `![` becomes `!\[`
- [x] visual editor doesn't distinguish between newline sequence (literal `\n`) and line break
- [x] html entities are decoded by the visual editor (not just visually, but in the source)
- [x] line break entities (trailing backslash) are visible in visual editor
- [x] markdown entities within code blocks should not be escaped
- [x] markdown entities within raw html should not be escaped
(trailing double space or backslash)
- [x] inactive markdown entities (like a literal asterisk or backslash) are unnecessarily escaped

#### UX issues
- [x] clicking below a trailing shortcode block in the visual editor should set focus in a paragraph
- [x] trim trailing empty paragraphs during remark -> markdown

#### Parsing issues
- [x] standalone URLs in markdown are converted to links in the visual editor
- [x] links containing images break the visual editor when attempting to backspace them
    (addressed here: https://github.com/ianstormtaylor/slate/pull/1202)
- [x] what do we do with footnotes?

#### Related issues
- [x] https://github.com/netlify/netlify-cms/issues/605 - can't edit data files
- [x] breaks sibling field updates when nested in a list as of https://github.com/netlify/netlify-cms/commit/c25a7c4abb76ed3ac041b18b86768fe94b70206f

### PR DESCRIPTION
**Pull request by erquhart**
Bugfixes and improvements for the new Slate editor. Continuation of #629. Closes #575.

**NOTE: merge after #638**

### CODE REVIEW

File: src/components/Widgets/Markdown/serializers/index.js [hunk 1]
```diff
@@ -118,27 +118,32 @@ export const remarkToMarkdown = obj => {
...
     strong: '*',
     rule: '-',
   };
 
   /**
-   * Escape markdown entities found in text and html nodes within the MDAST.
+   * Transform the MDAST with plugins.
    */
-  const escapedMdast = unified()
+  const processedMdast = unified()
```
Comment (tech4him1) [inline src/components/Widgets/Markdown/serializers/index.js]: 🎉 This makes a lot more sense.

File: src/components/Widgets/Markdown/serializers/index.js [hunk 2]
```diff
@@ -118,27 +118,32 @@ export const remarkToMarkdown = obj => {
...
     .use(remarkToMarkdownPlugin, remarkToMarkdownPluginOpts)
     .use(remarkAllowAllText)
-    .stringify(escapedMdast);
+    .stringify(processedMdast);
 
-  return markdown;
+  /**
+   * Return markdown with trailing whitespace removed.
+   */
+  return trimEnd(markdown);
```
Comment (tech4him1) [inline src/components/Widgets/Markdown/serializers/index.js]: This fixes at least two annoying bugs I've been having, thank you!

File: src/components/Widgets/Markdown/serializers/remarkEscapeMarkdownEntities.js [hunk 3]
```diff
@@ -248,6 +248,12 @@ function escape(delim) {
  */
 export default function remarkEscapeMarkdownEntities() {
   const transform = (node, index) => {
+    /**
+     * Shortcode nodes will intentionally inject markdown entities in text node
+     * children not be escaped.
```
Comment (tech4him1) [inline src/components/Widgets/Markdown/serializers/remarkEscapeMarkdownEntities.js]: Is there a typo here, or do I just not understand what it means?

Comment (erquhart) [reply src/components/Widgets/Markdown/serializers/remarkEscapeMarkdownEntities.js]: Nope, that's a typo.

**Comment by tech4him1** [comment]
If an undo (Ctrl+Z) causes a block to be removed (shift to another type of block), the editor loses focus.

Steps to Reproduce:
1. Use one of the files in the example (I am using post 10).
2. Add a new line of text in the visual editor.
3. Convert it to a different kind of block (header, code block, etc.).
4. Do an undo (Ctrl+Z). The editor will lose focus.

**Comment by tech4him1** [comment]
A bug related to the one above:

If a block is removed with (Ctrl+Z), the editor loses focus. Then, if you click in the editor so that it regains focus, you will need to press undo (Ctrl+Z) *twice* to undo the next thing in the undo history.

**Comment by erquhart** [comment]
Yeah, this is actually similar to the first bug you found - replacing the node that the cursor is in is the cause of all of them.

File: src/components/Widgets/Markdown/serializers/slateRemark.js [hunk 4]
```diff
@@ -435,7 +435,9 @@ function convertNode(node, children, shortcodePlugins) {
      * value and the "lang" data property to the new MDAST node.
      */
     case 'code': {
-      const value = get(node.nodes, [0, 'text']);
+      const value = flatMap(node.nodes, child => {
+        return flatMap(child.ranges, 'text');
```
[the remark this answers is not in the document]
Comment (erquhart) [reply src/components/Widgets/Markdown/serializers/slateRemark.js]: `ranges` is always an array.

[the remark this answers is not in the document]
Comment (Benaiah) [reply src/components/Widgets/Markdown/serializers/slateRemark.js]: What I was misunderstanding was that a string passed to the `iteratee` argument of `flatMap` will be treated as an argument to an implicit `properties` call, which makes this equivalent to something like `flatMap(get('text'), child.ranges)` in lodash/fp or ramda. LGTM

**Pull request merged**